cloudify_resources/tasks.py

```diff
@@ -113,8 +135,124 @@ def reserve_list_item(ctx, **kwargs):
     ))
 
 
+def _reserve_shared_list_item(ctx, **kwargs):
+    workflow_id = 'execute_operation'
+    reserve_params = {
+        'operation': 'cloudify.interfaces.operations.reserve',
+        'allow_kwargs_override': True,
+        'operation_kwargs': {
+            'reservation_id': ctx.source.instance.id
+        }
+    }
+    execute_workflow(workflow_id, reserve_params)
+
+    # Cloudify client setup
+    client_config = _get_target_shared_resource_client()
+    if client_config:
+        http_client = CloudifyClient(**client_config)
+    else:
+        http_client = manager.get_rest_client()
+
+    resources_list_node_id = kwargs.get('resources_list_node_id', None)
+    target_deployment_id = (ctx.target.node
+                            .properties['resource_config']
+                            ['deployment']['id'])
+
+    resources_list_instance = None
+
+    if resources_list_node_id:
+        resources_list_instance = http_client.node_instances.list(
+            deployment_id=target_deployment_id,
+            node_id=resources_list_node_id)[0]
+    else:
+        # if resources_list_node_id is not specified, first matching node
+        # will be used
+        for node in http_client.nodes.list(deployment_id=target_deployment_id):
+            if node.type == 'cloudify.nodes.resources.List':
+                resources_list_instance = http_client.node_instances.list(
+                    deployment_id=target_deployment_id,
+                    node_id=node.id)[0]
+                break
+
+    ctx.source.instance.runtime_properties[SINGLE_RESERVATION_PROPERTY] = \
+        resources_list_instance.runtime_properties.get(
+            RESERVATIONS_PROPERTY).get(ctx.source.instance.id)
+
+    ctx.logger.debug('Reservation successful: {0}\
+            \nLeft resources: {1}\
+            \nReservations: {2}'.format(
+        ctx.source.instance.runtime_properties.get(
+            SINGLE_RESERVATION_PROPERTY),
+        resources_list_instance.runtime_properties.get(
+            FREE_RESOURCES_LIST_PROPERTY),
+        resources_list_instance.runtime_properties.get(
+            RESERVATIONS_PROPERTY)
+    ))
+
+
+def _reserve_list_item(ctx, **kwargs):
+    if not ctx.instance.runtime_properties.get(
+            FREE_RESOURCES_LIST_PROPERTY, []):
+        raise NonRecoverableError(
+            'Reservation has failed, because there are no \
+            available resources right now.')
+
+    reservation_id = kwargs.get('reservation_id', None)
+    ctx.instance.refresh()
+
+    free_resources = \
+        ctx.instance.runtime_properties.get(
+            FREE_RESOURCES_LIST_PROPERTY)
+    reservations = \
+        ctx.instance.runtime_properties.get(
+            RESERVATIONS_PROPERTY, {})
+
+    if not reservation_id:
+        reservation_id = id_generator(size=8)
+        if reservation_id in reservations:
+            raise RecoverableError('reservation_id already taken.')
+    else:
```

Comment:
@jrzeszutek no need for the else just make the above code outside the if

Reply:
Done, thx @ahmadiesa-abu